### PR TITLE
feat: setTransactionRetentionDuration

### PR DESCRIPTION
**Pull request by vsmanish1772**
feat: add SetTransaction retention duration filtering

## What changes are proposed in this pull request?

This PR implements the `delta.setTransactionRetentionDuration` table property in delta-kernel-rs, which allows filtering out expired transaction identifiers based on their `lastUpdated` timestamp. This feature helps prevent unbounded growth of transaction logs and reduces checkpoint file sizes.

### Key changes:
1. **Transaction filtering during log replay**: Added retention timestamp filtering to `SetTransactionVisitor` to skip transactions older than the retention duration
2. **Checkpoint writing**: Updated `CheckpointVisitor` to exclude expired transactions when creating checkpoints
3. **Backward compatibility**: Transactions without `lastUpdated` field are preserved forever (never expire)
4. **Consistent behavior**: The same retention logic is applied in both checkpoint writing and batch transaction reading operations

### Implementation details:
- Modified `SetTransactionVisitor` to accept an optional retention timestamp
- Updated `CheckpointLogReplayProcessor` to calculate and propagate transaction retention timestamps
- Enhanced `CheckpointVisitor` to filter transactions based on retention during checkpoint creation
- Single transaction lookups (`get_app_id_version`) intentionally do filter to 
- Batch operations (future `get_all_app_id_versions`, checkpointing) do filter expired transactions

## How was this change tested?

Added comprehensive unit tests covering:

1. **Checkpoint writing tests** (`checkpoint::log_replay::tests`):
   - `test_checkpoint_visitor_txn_retention`: Verifies individual transaction filtering with various timestamp scenarios
   - `test_checkpoint_actions_iter_with_txn_retention`: Tests transaction filtering across multiple batches during checkpoint creation

2. **Transaction reading tests** (`actions::set_transaction::tests`):
   - `test_visitor_retention_with_null_last_updated`: Ensures transactions without `lastUpdated` are never filtered
   - Modified existing tests to support the new retention timestamp parameter

3. **Test scenarios covered**:
   - Transactions with `lastUpdated <= retention_timestamp` are filtered out
   - Transactions with `lastUpdated = None` are always kept (backward compatibility)
   - Boundary condition: transactions exactly at retention timestamp are filtered
   - Cross-batch behavior in checkpoint processing
   - Verification that filtered transactions are excluded from checkpoint files

All tests pass with `cargo t --all-features --all-targets`.

### CODE REVIEW

**Comment by vsmanish1772** [comment]
I think I should not pass optional retention_ts. I should set a default value as a fall back option. Thats still TBD in this pr

File: kernel/src/actions/set_transaction.rs [hunk 1]
```diff
@@ -165,4 +175,47 @@ mod tests {
...
+        let snapshot = table.snapshot(&engine, None).unwrap();
+        let log_segment = snapshot.log_segment();
+
+        // Test with no retention (should get all transactions)
+        let all_txns = SetTransactionScanner::get_all(log_segment, &engine, None).unwrap();
+        assert_eq!(all_txns.len(), 2);
+
+        // Test with retention that filters out old transactions
+        // Assuming the test data has transactions with different timestamps
+        let retention_timestamp = Some(i64::MAX); // Very recent timestamp
```
Comment (scovich) [inline kernel/src/actions/set_transaction.rs]: nit: that's a far future timestamp, not a very recent one?

File: kernel/src/actions/visitors.rs [hunk 2]
```diff
@@ -301,14 +301,18 @@ pub(crate) type SetTransactionMap = HashMap<String, SetTransaction>;
 pub(crate) struct SetTransactionVisitor {
     pub(crate) set_transactions: SetTransactionMap,
     pub(crate) application_id: Option<String>,
+    /// Minimum timestamp for transaction retention. Transactions with last_updated
+    /// older than this timestamp will be filtered out. None means no filtering.
+    retention_timestamp: Option<i64>,
```
Comment (scovich) [inline kernel/src/actions/visitors.rs]: Probably clearer to call it `expiration_timestamp` everywhere?
(especially since the filtering seems to use `<=` rather than `<`, contrary to this doc comment)

File: kernel/src/actions/visitors.rs [hunk 3]
```diff
@@ -353,6 +357,17 @@ impl RowVisitor for SetTransactionVisitor {
...
+                    // If last_updated is None, the transaction never expires
+                    if let Some(retention_ts) = self.retention_timestamp {
+                        if let Some(last_updated) = txn.last_updated {
+                            if last_updated <= retention_ts {
+                                // Transaction is old, skip it
+                                continue;
+                            }
+                        }
+                        // If last_updated is None, we keep the transaction (never expires)
+                    }
```
Comment (scovich) [inline kernel/src/actions/visitors.rs]:
```suggestion
                    match self.expiration_timestamp.zip(txn.last_updated) {
                        Some(expire_ts, last_updated) if last_updated <= expire_ts => continue,
                        _ => (),
                    }
```

File: kernel/src/checkpoint/log_replay.rs [hunk 4]
```diff
@@ -361,12 +372,28 @@ impl CheckpointVisitor<'_> {
...
+        getter: &[&'a dyn GetData<'a>],
+    ) -> DeltaResult<bool> {
         // Check for txn field
-        let Some(app_id) = getter.get_str(i, "txn.appId")? else {
+        let Some(app_id) = getter[11].get_str(i, "txn.appId")? else {
             return Ok(false); // Not a txn action
         };
 
+        // Check retention if last_updated is present
+        if let Some(retention_ts) = self.txn_retention_timestamp {
```
Comment (scovich) [inline kernel/src/checkpoint/log_replay.rs]: aside: There's a lot of redundancy here with the `SetTransactionVisitor` above. But it's just different enough I don't know what we could factor out.

File: kernel/src/checkpoint/log_replay.rs [hunk 5]
```diff
@@ -396,7 +423,7 @@ impl CheckpointVisitor<'_> {
         // The `||` operator short-circuits the evaluation, so if any of the checks return true,
         // the rest will not be evaluated.
         let is_valid = self.check_file_action(i, getters)?
-            || self.check_txn_action(i, getters[11])?
+            || self.check_txn_action(i, getters)?
```
Comment (scovich) [inline kernel/src/checkpoint/log_replay.rs:426]: aside: This existing code has sub-optimal short circuiting, see https://github.com/delta-io/delta-kernel-rs/issues/1014.

File: kernel/src/checkpoint/log_replay.rs [hunk 6]
```diff
@@ -504,6 +532,7 @@ mod tests {
             false,
             false,
             &mut seen_txns,
+            Some(0),
```
Comment (scovich) [inline kernel/src/checkpoint/log_replay.rs]: I suspect `None` would be more accurate in these tests, since they're not exercising app id expiration?

File: kernel/src/snapshot.rs [hunk 7]
```diff
@@ -363,6 +370,30 @@ impl Snapshot {
 
         domain_metadata_configuration(self.log_segment(), domain, engine)
     }
+
+    /// Calculate retention timestamp for transactions
+    fn calculate_transaction_retention_timestamp(&self) -> DeltaResult<Option<i64>> {
```
Comment (scovich) [inline kernel/src/snapshot.rs]: This logic is almost 100% duplicated from above. Can we make it an associated function that takes `&TableProperties` as input, which both call sites can just invoke?

Also, timestamp calculations in general have lots of duplicated logic, see https://github.com/delta-io/delta-kernel-rs/issues/877.

File: kernel/src/actions/visitors.rs (hunk 2 above)
Comment (vsmanish1772) [reply kernel/src/actions/visitors.rs]: That makes sense because that timestamp says when transactions expire. I have made changes to filtering and probably missed the doc comment. il make it clear. Thanks

File: kernel/src/actions/set_transaction.rs (hunk 1 above)
Comment (vsmanish1772) [reply kernel/src/actions/set_transaction.rs]: I have misjudged the data. I have updated the test with new test data.

File: kernel/src/snapshot.rs (hunk 7 above)
Comment (vsmanish1772) [reply kernel/src/snapshot.rs]: you mean its duplicated code from `deleted_file_retention_timestamp` method in CheckpointWriter? if so, I agree. I have copied that code. Initially, I wanted to create a shared function for both the uses cases but was waiting for some reviews on it.

Comment (vsmanish1772) [reply kernel/src/snapshot.rs]: Agreed. I initially wanted to create a utility shared function but wanted for some opinions. Il see what I can do with 
#877

File: kernel/src/utils.rs [hunk 8]
```diff
@@ -11,6 +14,29 @@ macro_rules! require {
...
+/// Calculates the transaction retention timestamp based on table properties.
+/// Returns None if set_transaction_retention_duration is not set.
+pub(crate) fn calculate_transaction_expiration_timestamp(
+    table_properties: &TableProperties,
+) -> DeltaResult<Option<i64>> {
+    match table_properties.set_transaction_retention_duration {
+        Some(duration) => {
+            let now = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .map_err(|e| Error::generic(format!("Failed to get current time: {}", e)))?;
```
Comment (scovich) [inline kernel/src/utils.rs]: Mixing results with options is tricky... does this help?
```rust
table_properties
    .set_transaction_retention_duration
    .map(|duration| {
          ...
        let now_ms = i64::try_from(...).map_err(|_| ...)?;
          ...
        Ok(now_ms - expiration_ms)
    })
    .transpose()
```

Comment (vsmanish1772) [reply kernel/src/utils.rs]: Ah I see. using map().transpose()  is much cleaner for handling the `Option<Result>` pattern. I wasn't familiar with the `transpose()` method, but now I see how it elegantly flips `Option<Result<T>>` to `Result<Option<T>>`, which is exactly what we need here.

File: kernel/src/checkpoint/log_replay.rs [hunk 9]
```diff
@@ -58,6 +58,8 @@ pub(crate) struct CheckpointLogReplayProcessor {
     seen_txns: HashSet<String>,
     /// Minimum timestamp for file retention, used for filtering expired tombstones.
     minimum_file_retention_timestamp: i64,
+    /// Transaction retention timestamp for filtering old transactions
```
Comment (scovich) [inline kernel/src/checkpoint/log_replay.rs]:
```suggestion
    /// Transaction expiration timestamp for filtering old transactions
```
(several doc comments that still mention "retention" instead of "expiration"?)

File: kernel/src/checkpoint/mod.rs [hunk 10]
```diff
@@ -222,6 +223,10 @@ impl CheckpointWriter {
 
         Ok(Self { snapshot, version })
     }
+
+    fn get_transaction_retention_timestamp(&self) -> DeltaResult<Option<i64>> {
```
Comment (scovich) [inline kernel/src/checkpoint/mod.rs]: retention? or expiration?

File: kernel/src/checkpoint/log_replay.rs [hunk 11]
```diff
@@ -361,12 +372,28 @@ impl CheckpointVisitor<'_> {
...
-        let Some(app_id) = getter.get_str(i, "txn.appId")? else {
+        let Some(app_id) = getter[11].get_str(i, "txn.appId")? else {
             return Ok(false); // Not a txn action
         };
 
+        // Check retention if last_updated is present
+        if let Some(retention_ts) = self.txn_expiration_timestamp {
+            if let Some(last_updated) = getter[12].get_opt(i, "txn.lastUpdated")? {
+                let last_updated: i64 = last_updated;
+                if last_updated <= retention_ts {
```
Comment (scovich) [inline kernel/src/checkpoint/log_replay.rs]: Could also do:
```suggestion
                if i64::le(&last_updated, &retention_ts) {
```
(not sure it's actually better tho)

Comment (vsmanish1772) [reply kernel/src/checkpoint/log_replay.rs]: Umm, coming from python background I would easily understand the code `if last_updated <= retention_ts`  at least from code readability.

Comment (vsmanish1772) [reply kernel/src/checkpoint/log_replay.rs]: Sure I can take this suggestion

File: kernel/src/actions/set_transaction.rs [hunk 12]
```diff
@@ -91,8 +99,10 @@ mod tests {
 
     use super::*;
     use crate::engine::sync::SyncEngine;
+    use crate::utils::test_utils::parse_json_batch;
     use crate::Table;
 
+    use arrow_55::array::StringArray;
```
Comment (zachschuermann) [inline kernel/src/actions/set_transaction.rs]:
```suggestion
    use crate::arrow::array::StringArray;
```

File: kernel/tests/data/app-txn-with-last-updated/part-00001-2e4c9beb-884e-4749-aaee-02a1a4f05750-c000.snappy.parquet
Comment (zachschuermann) [inline kernel/tests/data/app-txn-with-last-updated/part-00001-2e4c9beb-884e-4749-aaee-02a1a4f05750-c000.snappy.parquet:1]: do we need any data files in the test table?

Comment (vsmanish1772) [reply kernel/tests/data/app-txn-with-last-updated/part-00001-2e4c9beb-884e-4749-aaee-02a1a4f05750-c000.snappy.parquet:1]: probably not. I just made it consistent with other test tables. I will delete them in next commit if thats preferable

**Comment by vsmanish1772** [comment]
@zachschuermann do you anything from my side to get this approved ?

**Comment by zachschuermann** [comment]
> @zachschuermann do you anything from my side to get this approved ?

let's add that `#[allow(clippy::too_many_arguments)]` to get tests all green then i'll take another pass as soon as I can! (currently at a conference with limited availability)

File: kernel/src/checkpoint/log_replay.rs [hunk 13]
```diff
@@ -205,9 +212,12 @@ pub(crate) struct CheckpointVisitor<'seen> {
     // Set of transaction IDs to deduplicate by appId
     // This set has O(N) memory usage where N = number of txn actions with unique appIds
     seen_txns: &'seen mut HashSet<String>,
+    /// Transaction expiration timestamp for filtering old transactions
+    txn_expiration_timestamp: Option<i64>,
 }
 
 #[allow(unused)]
+#[allow(clippy::too_many_arguments)]
```
Comment (scovich) [inline kernel/src/checkpoint/log_replay.rs]: Probably better to add it to the `new` method that actually has the problem? 
Or was clippy flagging multiple methods?

Comment (vsmanish1772) [reply kernel/src/checkpoint/log_replay.rs]: Agreed. It was specifically on `new` method

**Comment by vsmanish1772** [comment]
> > @zachschuermann do you anything from my side to get this approved ?
> 
> let's add that `#[allow(clippy::too_many_arguments)]` to get tests all green then i'll take another pass as soon as I can! (currently at a conference with limited availability)

I have addressed the comment and is ready for another pass @zachschuermann

File: kernel/src/checkpoint/log_replay.rs [hunk 14]
```diff
@@ -361,12 +373,27 @@ impl CheckpointVisitor<'_> {
     /// Returns Ok(true) if the row contains a valid txn action.
     /// Returns Ok(false) if the row doesn't contain a txn action or is a duplicate.
```
Comment (zachschuermann) [inline kernel/src/checkpoint/log_replay.rs]: nit: can add "or is expired"

File: kernel/src/checkpoint/log_replay.rs [hunk 15]
```diff
@@ -361,12 +373,27 @@ impl CheckpointVisitor<'_> {
...
         // Check for txn field
-        let Some(app_id) = getter.get_str(i, "txn.appId")? else {
+        let Some(app_id) = getter[11].get_str(i, "txn.appId")? else {
             return Ok(false); // Not a txn action
         };
 
+        // Check retention if last_updated is present
+        if let Some(retention_ts) = self.txn_expiration_timestamp {
+            if let Some(last_updated) = getter[12].get_opt(i, "txn.lastUpdated")? {
+                if i64::le(&last_updated, &retention_ts) {
```
Comment (zachschuermann) [inline kernel/src/checkpoint/log_replay.rs:389]: nit: why not `last_update <= retention_ts`? casting issue?

**Review by zachschuermann — APPROVED** [review]
LGTM just a couple quick things! and can we add docs on this behavior to our overall `get_app_id_version` API - perhaps just a little doc comment that says we filter based on the table property + lastUpdated

sorry for the delay on the review, thanks for the contribution!!

File: kernel/src/checkpoint/log_replay.rs (hunk 15 above)
Comment (vsmanish1772) [reply kernel/src/checkpoint/log_replay.rs:389]: I have applied as suggestion from Ryan not any particular reason. initially I have implemented as you mentioned but I thought Ryan's suggestion was Rust way of doing things (I'm completely from python background)
